src/clients/Deluge.ts

```diff
@@ -0,0 +1,236 @@
+	private delugeCookie = "";
+	private delugeWebUrl: URL;
+	private delugeLabel: string;
+	private enLabeled: Promise<boolean>;
+
+	constructor() {
+		this.delugeWebUrl = new URL(`${getRuntimeConfig().delugeWebUrl}`);
+		this.delugeLabel = getRuntimeConfig().dataCategory;
+	}
+
+	/**
+	 * validates the login and host for deluge webui
+	 */
+	async validateConfig(): Promise<void> {
+		const url = new URL(this.delugeWebUrl);
+		if (url.username && !url.password) {
+			throw new CrossSeedError(
+				"you need to define a password in the delugeWebUrl. (eg: http://:<PASSWORD>@localhost:8112)"
+			);
+		}
+		const result = await this.authenticate();
+		if (!result) {
+			logger.verbose({
+				label: Label.DELUGE,
+				message: "failed to authenticate the deluge daemon.",
+			});
+			throw new CrossSeedError(
+				`failed to authenticate the deluge daemon.`
+			);
+		} else {
+			this.enLabeled = this.labelEnabled();
+		}
+	}
+
+	/**
+	 * connects and authenticates to the webui
+	 */
+	private async authenticate(): Promise<boolean> {
+		if (!this.loggedIn) {
+			let response = await this.call({
+				params: [this.delugeWebUrl.password],
+				method: "auth.login",
+			});
+			if (response === null) {
+				throw new CrossSeedError(
+					`Could not reach deluge: ${this.delugeWebUrl.origin}`
+				);
+			}
+			if (
+				response &&
+				response.headers &&
+				response["data"]["result"] &&
+				response.headers["set-cookie"]
+			) {
+				this.delugeCookie =
+					response.headers["set-cookie"][0].split(";")[0];
+				this.loggedIn = true;
+			}
+		} else {
+			let response = await this.call({
+				params: [this.delugeCookie],
+				method: "auth.check_session",
+			});
+			this.loggedIn = !!response;
+		}
+		return this.loggedIn;
+	}
+
+	/**
+	 * authenticates and sends JSON-RPC calls to deluge
+	 */
+	private async sendCall(method: string, params: any[]) {
+		let response;
+		if (!this.loggedIn) {
+			await this.authenticate();
+			response = await this.call({
+				params,
+				method,
+			});
+		} else
+			response = await this.call({
+				params: params,
+				method,
+			});
+
+		return response && response.data && response.data.error === null
+			? response.data
+			: response.data.error;
+	}
+
+	private async call(body: any): Promise<AxiosResponse | null> {
+		return new Promise((resolve) => {
+			body.id = ++this.msgId;
+			if (this.msgId > 1024) this.msgId = 0;
+			axios
+				.post(
+					this.delugeWebUrl.origin + this.delugeWebUrl.pathname,
+					body,
+					{
+						headers: {
+							"Content-Type": "application/json",
+							Cookie: this.delugeCookie,
+						},
+					}
+				)
+				.then((data) => {
+					resolve(data);
+				})
+				.catch(() => resolve(null));
+		});
+	}
+
+	/**
+	 * lists plugins and adds/sets labels
+	 * returns true if successful.
+	 */
+	private async labelEnabled() {
+		const enabledLabels = await this.sendCall(
+			"core.get_enabled_plugins",
+			[]
+		);
+		return enabledLabels["result"].includes("Label");
```

Comment:
nit: anywhere we do `a["b"]`, we can also do `a.b` which is preferable. We can make similar changes in other places in this PR as well

```suggestion
		return enabledLabels.result.includes("Label");
```

Reply:
Will apply as suggested.